### PR TITLE
Add group block enumeration method

### PR DESCRIPTION
**Pull request by prachigauriar**
@macdrevx @jnjosh Please review.

Adds a block enumerator method that returns a dictionary that groups the elements in a collection by their return value from a block. For example:

```objc
NSArray *iOSDevelopers = @[ @"Ameir", @"Andrew", @"Duncan", @"Jill", @"Josh", @"Kevin", @"Prachi", @"Steve", @"Tom" ];

NSDictionary *groups = [iOSDevelopers twt_groupWithBlock:^id<NSCopying>(NSString *name) {
    return @(name.length);
}];
```

The resulting `groups` dictionary would look like:

```objc
@{ @3 : @[ @"Tom" ],
   @4 : @[ @"Jill", @"Josh" ],
   @5 : @[ @"Ameir", @"Kevin", @"Steve" ],
   @6 : @[ @"Andrew", @"Duncan", @"Prachi" ] }
```

### CODE REVIEW

File: Foundation/Block Enumeration/TWTBlockEnumeration.h [hunk 1]
```diff
@@ -72,18 +80,31 @@ typedef id (^TWTBlockEnumerationInjectBlock)(id memo, id element);
  @result A new instance of the collection with the results of invoking the block on each element of
     the original collection.
  */
-- (instancetype)twt_collectWithBlock:(TWTBlockEnumerationCollectBlock)block;
+- (id)twt_collectWithBlock:(TWTBlockEnumerationCollectBlock)block;
```
Comment (prachigauriar) [inline Foundation/Block Enumeration/TWTBlockEnumeration.h]: These changes are because the methods don’t return `instancetype` for `NSEnumerator`.

Comment (jnjosh) [inline Foundation/Block Enumeration/TWTBlockEnumeration.h]: Good catch on that!

**Comment by jnjosh** [comment]
Looks good to me. I feel like we should add the equivalent to flatMap too… because, reasons…

**Comment by kconner** [comment]
I'd want this block to receive the object's array index, too.

**Comment by prachigauriar** [comment]
@kconner It might not be an array. This works for sets, ordered sets, and dictionaries too.

**Pull request merged**